### ISSUE TITLE
Develop more coherent UI architecture

### ISSUE THREAD

**Issue by aeickhoff** [citybound/citybound#91]
Currently messy code and uncertainties in an increasing number of areas hint at a missing coherent UI logic/architecture. In this meta issue I'll collect my own thoughts and developments towards an actual architecture.

**Comment by aeickhoff** [comment]
Currently collected topics:

Collected topics:
- `delta_time` for purely graphical purposes, for example camera movement
- different "modes" of capturing input: ButtonInteractable, MouseInteractable and TextInteractable ?
- how to organise in general different UI modes/pages, components, renderables...

**Comment by kingoflolz** [comment]
On the different modes of capturing input, I think I know an elegant way of handling that. A single enum stores all the input modes, e.g. one variant for mouse (which would contain things like DragStart etc), one variant for text input (not used yet, but should handle things like clipboard) and one variant for bindable buttons, which essentially allows interactables to register button presses and mapping to semantic meaning on the fly, recieving a String to match against to trigger their own actions.

If an intreactable wants to "sub in" to an input type, they just match against that instead of doing `_=>()` on them

**Comment by aeickhoff** [comment]
Sounds good! Could you elaborate on / invent how the registering of semantic buttons would work?

**Comment by kingoflolz** [comment]
It would be somewhat similar to registering actors, there would be a global InputDispatcher or something which you use register(String, vec<keyOrButton>) on, and internally it will keep track of everything and it will send the interactable that String if any of the keyOrButtons are held down.

**Comment by aeickhoff** [comment]
All of that should be implementable in the current code, independently of any kind of higher-level UI refactoring, right?

**Comment by kingoflolz** [comment]
It would require factoring out the current UI into an intreactable and then adding InputDispatcher. Not sure if that would be higher-level UI refactoring.

**Comment by aeickhoff** [comment]
Why can't you do the dispatching directly in UI and just change the messaging interface UI has with interactables?

**Comment by kingoflolz** [comment]
Thats possible, but it would be special casing for camera movements, so you will need two sets on handling, one for "local" actions such as camera movements and one for "remote" actions which is sent to another interactable.

**Comment by aeickhoff** [comment]
Wouldn't it be nicer anyways to pull camera movements into an interactable? (relevant for #85)

**Comment by kingoflolz** [comment]
Yeah, think so. Maybe I'll wait until @Herbstein is done with #85 before I start to tackle this (otherwise merge conflicts galore).

**Comment by aeickhoff** [comment]
As far as I understood, @Herbstein is actually blocked by this very issue and waiting.

**Comment by kingoflolz** [comment]
Hrm, ok. In that case I guess I'll start on that.

**Comment by Herbstein** [comment]
I'm indeed stuck on getting mouse movement where I want it without doing any major refactoring.

**Comment by kingoflolz** [comment]
```
   --> src/core/ui.rs:127:10
    |
127 | #[derive(Compact, Clone)]
    |          ^^^^^^^
    |
    = help: message: Only tuple enum variants supported so far

error: Could not compile `citybound`.
```
D:

**Comment by Herbstein** [comment]
Isn't that because userdefined traits can't be used in `#[derive()]`? The rust-book has a [list](http://rustbyexample.com/trait/derive.html) of the traits that can be used. All others must be explicitly implemented.

There's been talk of adding support for deriving userdefined traits and I think that's what the error message is referring to. They might've only implemented it for `tuple enum variants`.

**Comment by kingoflolz** [comment]
It is just a unimplemented part of Kay Macros, but I managed to find a work around.

**Comment by Herbstein** [comment]
Damn. Everytime I think I know what's wrong I miss the mark completely. I shouldn't be so quick to spew with "knowledge" :P

**Comment by kingoflolz** [comment]
Another question @aeickhoff, how do you send a string in a message?

**Comment by aeickhoff** [comment]
Not easily yet, I'll implement it today

**Comment by aeickhoff** [comment]
Superseded by #171
